File: examples/Vector2.php
<?php
/**
 * Created by IntelliJ IDEA.
 * User: Jack
 * Date: 4/11/18
 * Time: 10:37 PM
 */

$vec = new \pocketmine\math\Vector2(mt_rand(-99, 99) + (mt_rand(-99, 99) / 10), mt_rand(-99, 99) + (mt_rand(-99, 99) / 10));

echo $vec->getX() . PHP_EOL;
echo $vec->getY() . PHP_EOL;

echo $vec->getFloorX() . PHP_EOL;
echo $vec->getFloorY() . PHP_EOL;

echo $vec->add(0.5, 0.5) . PHP_EOL;
echo $vec->subtract(0.5, 0.5) . PHP_EOL;

echo $vec->ceil() . PHP_EOL;
echo $vec->floor() . PHP_EOL;
echo $vec->round() . PHP_EOL;
echo $vec->abs() . PHP_EOL;
echo $vec->multiply(2) . PHP_EOL;
echo $vec->divide(2) . PHP_EOL;

echo $vec->distance(0, 0) . PHP_EOL;
echo $vec->distanceSquared(0, 0) . PHP_EOL;

echo $vec->normalize() . PHP_EOL;

echo $vec->dot($vec->multiply(2)) . PHP_EOL;

echo $vec . PHP_EOL;

File: lib/Vector2.cpp
//
// Created by Jack Noordhuis on 11/11/18.
//

#include "Vector2.h"

#include <cmath>

int Vector2::getFloorX() {
    return (int) ::floor(x);
}

int Vector2::getFloorY() {
    return (int) ::floor(y);
}

Vector2 *Vector2::ceil() {
    return new Vector2((int) ::ceil(x), (int) ::ceil(y));
}

Vector2 *Vector2::floor() {
    return new Vector2((int) ::floor(x), (int) ::floor(y));
}

Vector2 *Vector2::round() {
    return new Vector2((int) ::round(x), (int) ::round(y));
}

Vector2 *Vector2::abs() {
    return new Vector2((int) ::abs(x), (int) ::abs(y));
}

double Vector2::distance(double x, double y) {
    return ::sqrt(distanceSquared(x, y));
}

double Vector2::distanceSquared(double x, double y) {
    return ::pow(this->x - x, 2) + ::pow(this->y - y, 2);
}

double Vector2::length() {
    return ::sqrt(lengthSquared());
}

Vector2 *Vector2::normalize() {
    double length = lengthSquared();
    if(length > 0) {
        return divide(::sqrt(length));
    }

    return new Vector2(0, 0);
}